shell/platform/darwin/macos/framework/Source/FlutterMouseCursorPlugin.mm

```diff
@@ -0,0 +1,171 @@
+  if (self != nil) {
+    _flutterViewController = viewController;
+    _channel = [FlutterMethodChannel methodChannelWithName:kMouseCursorChannel
+                                           binaryMessenger:viewController.engine.binaryMessenger
+                                                     codec:[FlutterStandardMethodCodec sharedInstance]];
+    __weak FlutterMouseCursorPlugin* weakSelf = self;
+    [_channel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
+      [weakSelf handleMethodCall:call result:result];
+    }];
+    _shapeObjects = [[NSMutableDictionary alloc] init];
+  }
+  return self;
+}
+
+#pragma mark - Private
+
+- (void)handleMethodCall:(nonnull FlutterMethodCall*)call result:(FlutterResult)result {
+  BOOL handled = YES;
+  NSString* method = call.method;
+  if ([method isEqualToString:kActivateSystemCursorMethod]) {
+    result([self activateSystemCursor:call.arguments]);
+  } else {
+    handled = NO;
+    NSLog(@"Unhandled mouse cursor method '%@'", method);
+    result(FlutterMethodNotImplemented);
+  }
+}
+
+- (id)activateSystemCursor:(nonnull NSDictionary*)arguments {
+  if (!arguments) {
+    return [FlutterError
+        errorWithCode:@"error"
+              message:@"Missing arguments"
+              details:@"Missing arguments while trying to activate system cursor"];
+  }
+  NSNumber* shapeCodeArg = arguments[kShapeCodeKey];
+  if (!shapeCodeArg) {
+    return [FlutterError
+        errorWithCode:@"error"
+              message:@"Missing argument"
+              details:@"Missing argument shapeCode while trying to activate system cursor"];
+  }
+  NSNumber* deviceArg = arguments[kDeviceKey];
+  if (!deviceArg) {
+    return [FlutterError
+        errorWithCode:@"error"
+              message:@"Missing argument"
+              details:@"Missing argument device while trying to activate system cursor"];
+  }
+  if ([shapeCodeArg intValue] == kSystemShapeNone) {
+    return [self hide];
+  }
+  NSCursor* shapeObject = [self resolveShapeCode:shapeCodeArg];
+  if (shapeObject == nil) {
+    // Unregistered shape. Return false to request fallback.
+    return @(NO);
+  }
+  return [self activateShapeObject:shapeObject];
+}
+
+- (nullable NSCursor*)resolveShapeCode:(NSNumber*)shape {
+  NSCursor *cachedObject = _shapeObjects[shape];
+  if (cachedObject)
+    return cachedObject;
+  NSCursor *systemObject = [FlutterMouseCursorPlugin resolveSystemShape:shape];
+  if (!systemObject)
+    return nil;
+  _shapeObjects[shape] = systemObject;
+  return systemObject;
+}
+
+- (id)activateShapeObject: (nonnull NSCursor *)targetShape {
+  if (targetShape == _shape && !_hidden)
+    return @(YES);
+  [targetShape set];
+  if (_hidden) {
+    [NSCursor unhide];
+  }
+  _shape = targetShape;
+  _hidden = NO;
+  return @(YES);
+}
+
+- (id)hide {
+  if (!_hidden) {
+    [NSCursor hide];
+  }
+  _hidden = YES;
+  return @(YES);
+}
+
+#pragma mark - Static
+
+// Does not handle kSystemShapeNone.
```

Comment:
The first sentence of the declaration comment should explain what it does. Then you can follow up with details like things it doesn't handle.